presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -458,16 +504,120 @@ else if (isCovariantParameterPosition(commonSuperTypeBase.get(), i)) {
         return Optional.of(new TypeSignature(commonSuperTypeBase.get(), typeParameters.build()));
     }
 
-    /*
-     * Given signature with unmatched parameters return filtered one.
+    /**
+     * In case of coercion between un-parametrized types (DOUBLE, BIGINT, ...)
+     * we only need to know whether coercion is allowed (true|false).
+     * But in case of coercion to the type with long parameters, parameters of the result type must be calculated.
+     * Such coercion rules must be implemented within this method.
+     *
+     * Examples:
+     *
+     * LONG, DECIMAL(x,y) -> DECIMAL(19,0)
```

Comment:
Maybe we can leave `canCoerce` and `getCommonSuperType` methods that only accept `Type`. And than add some kind of `canBind` and `getBindResult` that accept `Type` and `TypeSignature`?